<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Genotype simulation supported by *PhenotypeSimulator* • PhenotypeSimulator</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Genotype simulation supported by *PhenotypeSimulator*">
<meta property="og:description" content="PhenotypeSimulator">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-standard navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">PhenotypeSimulator</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home"></span>
     
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fas fa-file"></span>
     
    documentation
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/PhenotypeSimulator.html">Step-by-step phenotype simulation</a>
    </li>
    <li>
      <a href="../articles/Simulation-and-LinearModel.html">Application of PhenotypeSimulator for a power analysis study</a>
    </li>
    <li>
      <a href="../articles/Sample-scripts-external-genotype-simulation.html">Genotype simulation supported by PhenotypeSimulator</a>
    </li>
    <li>
      <a href="../articles/SimulationBasedOnExampleData.html">Phenotype simulation based on an example dataset</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="../news/index.html">
    <span class="fas fa-newspaper-o"></span>
     
    news
  </a>
</li>
<li>
  <a href="../reference/index.html">
    <span class="fas fa-code"></span>
     
    functions
  </a>
</li>
<li>
  <a href="https://github.com/HannahVMeyer/PhenotypeSimulator">
    <span class="fas fa-github fa-lg"></span>
     
    github
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="sample-scripts-external-genotype-simulation_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Genotype simulation supported by <em>PhenotypeSimulator</em>
</h1>
                        <h4 class="author">Hannah Meyer</h4>
            
            <h4 class="date">2021-07-14</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/HannahVMeyer/PhenotypeSimulator/blob/master/vignettes/sample-scripts-external-genotype-simulation.Rmd"><code>vignettes/sample-scripts-external-genotype-simulation.Rmd</code></a></small>
      <div class="hidden name"><code>sample-scripts-external-genotype-simulation.Rmd</code></div>

    </div>

    
    
<p>There are a number of different strategies to generate genotype data for genetic association studies.</p>
<ol style="list-style-type: decimal">
<li><p><strong>Simple bi-allelic SNPs without structure</strong>: In the most simple case and assuming bi-allelic SNPs, each SNP is simulated from a binomial distribution with two trials and probability equal to the given allele frequencies. This simple approach does not simulate any dependency between the genotypes as is observed with LD structure in the genome. Simple bi-allelic SNPs can be simulated with PhenotypeSimulator::simulateGenotypes or via PLINK <span class="citation">[1]</span>. For large datasets (~ 1 million SNPs), simulation via PLINK is recommended.</p></li>
<li><p><strong>Coalescent approaches</strong>: Coalescent methods simulate genealogical events backward in time. These events typically include the coalescence of two sequences into a single ancestral lineage, recombination within a sequence or migration between populations. A coalescent-based approach to simulate whole genome data is implemented in GENOME <span class="citation">[2]</span> and its output format is supported as an input format for <em>PhenotypeSimulator</em>.</p></li>
<li><p><strong>Forward-time approaches</strong>: Forward-time simulation methods evolve a population forward in time, subject to arbitrary genetic and demographic factors <span class="citation">[3]</span>. Several algorithms are available, many of which allow customisation by building the simulation scheme in R or python, hence the output format of the genotypes can be specified by the user. <em>PhenotypeSimulator</em> offers reading genotypes from delimited-files and as such, any genotypes generated by these programmes can be read (e.g. MetaSim <span class="citation">[4]</span> or simuPop <span class="citation">[5]</span>).</p></li>
<li><p><strong>Resampling approaches</strong>: Resampling-based approaches combine existing genotype data into the genotypes of the simulated samples, thereby retaining allele frequency and LD patterns <span class="citation">[6]</span>. A standard resampling-based approach that uses common genotype formats (common for the <a href="http://www.stats.ox.ac.uk/~marchini/software/gwas/file_format.html">oxford genetics format</a>) is Hapgen2 and an example usage is given below.</p></li>
</ol>
<p>Examples for genotype simulation via PLINK, GENOME and Hapgen2 are given below. The corresponding data can be found in the extdata folder.</p>
<div id="simple-bi-allelic-snps-without-structure-via-plink" class="section level2">
<h2 class="hasAnchor">
<a href="#simple-bi-allelic-snps-without-structure-via-plink" class="anchor"></a>Simple bi-allelic SNPs without structure via PLINK</h2>
<p>Download <a href="https://www.cog-genomics.org/plink/1.9/">PLINK</a> and create a folder for the PLINK output files. The example below uses PLINK to simulate 1000 SNPs, with allele frequencies between 0 and 1 for 100 controls and 0 cases. PhenotypeSimulator::readStandardGenotypes reads the resulting .bim, .fam, .bed files.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1"></a><span class="co"># Serves as output directory for simulation and parameter file</span></span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="fu">mkdir</span> -p ~/PhenotypeSimulator/inst/extdata/genotypes/plink</span>
<span id="cb1-3"><a href="#cb1-3"></a><span class="bu">cd</span> ~/PhenotypeSimulator/inst/extdata/genotypes/plink</span>
<span id="cb1-4"><a href="#cb1-4"></a></span>
<span id="cb1-5"><a href="#cb1-5"></a><span class="co"># Write paramter file to simulate 1000 SNPs, with allele frequencies between 0 </span></span>
<span id="cb1-6"><a href="#cb1-6"></a><span class="co"># and 1 for 100 controls and 0 cases </span></span>
<span id="cb1-7"><a href="#cb1-7"></a><span class="bu">echo</span> -e <span class="st">"1000\tSNP\t0.00\t1.00\t1.00\t1.00"</span> <span class="op">&gt;</span> plink_sim_par.txt</span>
<span id="cb1-8"><a href="#cb1-8"></a></span>
<span id="cb1-9"><a href="#cb1-9"></a><span class="ex">plink</span> --simulate plink_sim_par.txt \</span>
<span id="cb1-10"><a href="#cb1-10"></a>      --simulate-ncases 0 \</span>
<span id="cb1-11"><a href="#cb1-11"></a>      --simulate-ncontrols 100 \</span>
<span id="cb1-12"><a href="#cb1-12"></a>      --out genotypes_plink</span></code></pre></div>
</div>
<div id="coalescent-simulation-via-genome" class="section level2">
<h2 class="hasAnchor">
<a href="#coalescent-simulation-via-genome" class="anchor"></a>Coalescent simulation via Genome</h2>
<p>Download <a href="http://csg.sph.umich.edu/liang/genome/download.html">GENOME</a> and create a folder for the GENOME output files. The example below uses GENOME to simulate genetic data for a population comprised of three sub-populations with 30, 30 and 40 samples. The simulated POPULATION PROFILE, SNP positions and the genotypes are all saved in genotypes_genome.txt. PhenotypeSimulator::readStandardGenotypes reads genotype information by parsing this output file and extracting the samples information following the line ‘Samples:’</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1"></a><span class="fu">mkdir</span> -p ~/PhenotypeSimulator/inst/extdata/genotypes/genome</span>
<span id="cb2-2"><a href="#cb2-2"></a><span class="bu">cd</span> ~/PhenotypeSimulator/inst/extdata/genotypes/genome</span>
<span id="cb2-3"><a href="#cb2-3"></a></span>
<span id="cb2-4"><a href="#cb2-4"></a><span class="co">#  subpopulation with 30, 30 and 40 individuals each</span></span>
<span id="cb2-5"><a href="#cb2-5"></a><span class="ex">genome</span> -pop 3 30 30 40 <span class="op">&gt;</span> genotypes_genome.txt</span></code></pre></div>
</div>
<div id="resampling-based-simulation-via-hapgen2" class="section level2">
<h2 class="hasAnchor">
<a href="#resampling-based-simulation-via-hapgen2" class="anchor"></a>Resampling-based simulation via Hapgen2</h2>
<p>Download <a href="http://mathgen.stats.ox.ac.uk/genetics_software/hapgen/hapgen2.html">hapgen2</a>, <a href="http://mathgen.stats.ox.ac.uk/genetics_software/hapgen/download/example/hapgen2.example.tgz">hapgen example files</a> and the 1000Genomes data from the <a href="https://mathgen.stats.ox.ac.uk/impute/impute_v1.html#Using_IMPUTE_with_the_HapMap_Data">impute2 webpage</a> as starting point for the resampling-based genotype simulation with hapgen2. For formating of the 1000Genomes data, have a look at this <a href="https://hannahvmeyer.github.io/PhenotypeSimulator/articles/Simulation-and-LinearModel.html#genotype-simulation-via-hapgen2">vignette</a>. The example files contain example haplotypes (ex.haps), legend files (ex.leg) map (ex.map) and TagSNP files (ex.tags). The following examples simulates genotype data for 100 controls and 0 cases with 1000 SNPs from chromosome 1. PhenotypeSimulator::readStandardGenotypes reads the resulting .gen and .samples files.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1"></a><span class="co"># contains ex.leg file and serves as output directory</span></span>
<span id="cb3-2"><a href="#cb3-2"></a><span class="fu">mkdir</span> -p ~/PhenotypeSimulator/inst/extdata/genotypes/hapgen</span>
<span id="cb3-3"><a href="#cb3-3"></a><span class="bu">cd</span> ~/PhenotypeSimulator/inst/extdata/genotypes/hapgen</span>
<span id="cb3-4"><a href="#cb3-4"></a></span>
<span id="cb3-5"><a href="#cb3-5"></a><span class="co"># contains 1000Genomes haplotype data used for sampling</span></span>
<span id="cb3-6"><a href="#cb3-6"></a><span class="va">hapdir=</span>/path/to/CEU.0908.impute.files</span>
<span id="cb3-7"><a href="#cb3-7"></a></span>
<span id="cb3-8"><a href="#cb3-8"></a><span class="ex">hapgen2</span> -m <span class="va">$hapdir</span>/genetic_map_chr1_combined_b37.txt \</span>
<span id="cb3-9"><a href="#cb3-9"></a>        -l ex.leg \</span>
<span id="cb3-10"><a href="#cb3-10"></a>        -h <span class="va">$hapdir</span>/chr1.ceu_subset.hap \</span>
<span id="cb3-11"><a href="#cb3-11"></a>        -o genotypes_hapgen \</span>
<span id="cb3-12"><a href="#cb3-12"></a>        -n 100 0 \</span>
<span id="cb3-13"><a href="#cb3-13"></a>        -dl 45162 0 0 0 \</span>
<span id="cb3-14"><a href="#cb3-14"></a>        -no_haps_output</span></code></pre></div>
</div>
<div id="references" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<div id="refs" class="references">
<div id="ref-Chang2015">
<p>1. Chang CC, Chow CC, Tellier LC, Vattikuti S, Purcell SM, Lee JJ. Second-generation PLINK: rising to the challenge of larger and richer datasets. GigaScience. 2015;4: 7. doi:<a href="https://doi.org/10.1186/s13742-015-0047-8">10.1186/s13742-015-0047-8</a></p>
</div>
<div id="ref-Liang2007">
<p>2. Liang L, Zollner S, Abecasis GR. GENOME: a rapid coalescent-based whole genome simulator. Bioinformatics. 2007;23: 1565–1567. doi:<a href="https://doi.org/10.1093/bioinformatics/btm138">10.1093/bioinformatics/btm138</a></p>
</div>
<div id="ref-Peng2010">
<p>3. Peng B, Amos CI. Forward-time simulation of realistic samples for genome-wide association studies. BMC Bioinformatics. 2010;11: 442. doi:<a href="https://doi.org/10.1186/1471-2105-11-442">10.1186/1471-2105-11-442</a></p>
</div>
<div id="ref-Strand2002">
<p>4. Strand AE. Metasim 1.0: An individual-based environment for simulating population genetics of complex population dynamics. Molecular Ecology Notes. 2002;2: 373–376. doi:<a href="https://doi.org/10.1046/j.1471-8286.2002.00208.x">10.1046/j.1471-8286.2002.00208.x</a></p>
</div>
<div id="ref-Peng2005">
<p>5. Peng B, Kimmel M. simuPOP: a forward-time population genetics simulation environment. Bioinformatics. 2005;21: 3686–3687. doi:<a href="https://doi.org/10.1093/bioinformatics/bti584">10.1093/bioinformatics/bti584</a></p>
</div>
<div id="ref-Wright2007">
<p>6. Wright FA, Huang H, Guan X, Gamiel K, Jeffries C, Barry WT, et al. Simulating association studies: a data-based resampling method for candidate regions or whole genome scans. Bioinformatics. 2007;23: 2581–2588. doi:<a href="https://doi.org/10.1093/bioinformatics/btm386">10.1093/bioinformatics/btm386</a></p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Hannah Meyer.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
