<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Set simulation model. — setModel • PhenotypeSimulator</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  

<meta property="og:title" content="Set simulation model. — setModel" />
<meta property="og:description" content="Based on parameters provided, this function sets the name for the phenotype 
simulation. It carries out compatibiltiy checks of the specifie parameters 
and checks for any missing information." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-standard navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">PhenotypeSimulator</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home"></span>
     
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fas fa-file"></span>
     
    documentation
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/PhenotypeSimulator.html">Step-by-step phenotype simulation</a>
    </li>
    <li>
      <a href="../articles/Simulation-and-LinearModel.html">Application of PhenotypeSimulator for a power analysis study</a>
    </li>
    <li>
      <a href="../articles/Sample-scripts-external-genotype-simulation.html">Genotype simulation supported by PhenotypeSimulator</a>
    </li>
    <li>
      <a href="../articles/SimulationBasedOnExampleData.html">Phenotype simulation based on an example dataset</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="../news/index.html">
    <span class="fas fa-newspaper-o"></span>
     
    news
  </a>
</li>
<li>
  <a href="../reference/index.html">
    <span class="fas fa-code"></span>
     
    functions
  </a>
</li>
<li>
  <a href="https://github.com/HannahVMeyer/PhenotypeSimulator">
    <span class="fas fa-github fa-lg"></span>
     
    github
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Set simulation model.</h1>
    <small class="dont-index">Source: <a href='https://github.com/HannahVMeyer/PhenotypeSimulator/blob/master/R/createphenotypeFunctions.R'><code>R/createphenotypeFunctions.R</code></a></small>
    <div class="hidden name"><code>setModel.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Based on parameters provided, this function sets the name for the phenotype 
simulation. It carries out compatibiltiy checks of the specifie parameters 
and checks for any missing information.</p>
    </div>

    <pre class="usage"><span class='fu'>setModel</span><span class='op'>(</span>genVar <span class='op'>=</span> <span class='cn'>NULL</span>, h2s <span class='op'>=</span> <span class='cn'>NULL</span>, theta <span class='op'>=</span> <span class='fl'>0.8</span>, h2bg <span class='op'>=</span> <span class='cn'>NULL</span>,
  eta <span class='op'>=</span> <span class='fl'>0.8</span>, noiseVar <span class='op'>=</span> <span class='cn'>NULL</span>, delta <span class='op'>=</span> <span class='cn'>NULL</span>, gamma <span class='op'>=</span> <span class='fl'>0.8</span>,
  rho <span class='op'>=</span> <span class='cn'>NULL</span>, phi <span class='op'>=</span> <span class='cn'>NULL</span>, alpha <span class='op'>=</span> <span class='fl'>0.8</span>, pcorr <span class='op'>=</span> <span class='fl'>0.6</span>,
  pIndependentConfounders <span class='op'>=</span> <span class='fl'>0.4</span>, pTraitIndependentConfounders <span class='op'>=</span> <span class='fl'>0.2</span>,
  pIndependentGenetic <span class='op'>=</span> <span class='fl'>0.4</span>, pTraitIndependentGenetic <span class='op'>=</span> <span class='fl'>0.2</span>,
  proportionNonlinear <span class='op'>=</span> <span class='fl'>0</span>, cNrSNP <span class='op'>=</span> <span class='cn'>NULL</span>, NrConfounders <span class='op'>=</span> <span class='fl'>10</span>,
  verbose <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>genVar</th>
      <td><p>Total genetic variance [double].</p></td>
    </tr>
    <tr>
      <th>h2s</th>
      <td><p>Proportion [double] of variance of genetic variant effects.</p></td>
    </tr>
    <tr>
      <th>theta</th>
      <td><p>Proportion [double] of variance of shared genetic variant 
effects.</p></td>
    </tr>
    <tr>
      <th>h2bg</th>
      <td><p>Proportion [double] of variance of infinitesimal genetic effects
i.e. correlation introduced by sample kinship).</p></td>
    </tr>
    <tr>
      <th>eta</th>
      <td><p>Proportion [double] of variance of shared infinitesimal genetic 
effects.</p></td>
    </tr>
    <tr>
      <th>noiseVar</th>
      <td><p>Total noise variance [double].</p></td>
    </tr>
    <tr>
      <th>delta</th>
      <td><p>Proportion [double] of variance of non-genetic covariate effect.</p></td>
    </tr>
    <tr>
      <th>gamma</th>
      <td><p>Proportion [double] of variance of shared non-genetic covariate 
effects.</p></td>
    </tr>
    <tr>
      <th>rho</th>
      <td><p>Proportion [double] of variance of correlated noise effects.</p></td>
    </tr>
    <tr>
      <th>phi</th>
      <td><p>Proportion [double] of variance of observational noise effects.</p></td>
    </tr>
    <tr>
      <th>alpha</th>
      <td><p>Proportion [double] of variance of shared observational noise 
effect.</p></td>
    </tr>
    <tr>
      <th>pcorr</th>
      <td><p>Correlation [double] between phenotypes.</p></td>
    </tr>
    <tr>
      <th>pIndependentConfounders</th>
      <td><p>Proportion [double] of non-genetic covariate 
to have a trait-independent effect.</p></td>
    </tr>
    <tr>
      <th>pTraitIndependentConfounders</th>
      <td><p>Proportion [double] of traits influenced
by independent non-genetic covariate effects.</p></td>
    </tr>
    <tr>
      <th>pIndependentGenetic</th>
      <td><p>Proportion [double] of genetic variant effects to 
have a trait-independent fixed effect.</p></td>
    </tr>
    <tr>
      <th>pTraitIndependentGenetic</th>
      <td><p>Proportion [double] of traits influenced by 
independent genetic variant effects.</p></td>
    </tr>
    <tr>
      <th>proportionNonlinear</th>
      <td><p>[double] proportion of the phenotype to be non-
linear</p></td>
    </tr>
    <tr>
      <th>cNrSNP</th>
      <td><p>Number [integer] of causal SNPs; used as genetic variant 
effects.</p></td>
    </tr>
    <tr>
      <th>NrConfounders</th>
      <td><p>Number [integer] of non-genetic covariates; used as 
non-genetic covariate effects.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>[boolean]; If TRUE, progress info is printed to standard out.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Named list containing the genetic model (modelGenetic), the noise 
model (modelNoise) and the input parameters (h2s, h2bg, noiseVar, rho, delta, 
phi, gamma, theta, eta, alpha, pcorr, proportionNonlinear). Model options 
are: modelNoise: "noNoise", "noiseFixedOnly", "noiseBgOnly", 
"noiseCorrelatedOnly", "noiseFixedAndBg","noiseCorrelatedAndBg", 
"noiseFixedAndCorrelated", "noiseFixedAndBgAndCorrelated"
modelGenetic: "noGenetic","geneticBgOnly", "geneticFixedOnly",
"geneticFixedAndBg"</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'>#genetic fixed effects only</span>
<span class='va'>model</span> <span class='op'>&lt;-</span> <span class='fu'>setModel</span><span class='op'>(</span>genVar<span class='op'>=</span><span class='fl'>1</span>, h2s<span class='op'>=</span><span class='fl'>1</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>The total noise variance (noiseVar) is: 0</span></div><div class='output co'>#&gt; <span class='message'>The noise model is: noNoise</span></div><div class='output co'>#&gt; <span class='message'>The total genetic variance (genVar) is: 1</span></div><div class='output co'>#&gt; <span class='message'>The genetic model is: geneticFixedOnly</span></div><div class='output co'>#&gt; <span class='message'>Proportion of variance of genetic variant effects (h2s): 1</span></div><div class='output co'>#&gt; <span class='message'>Proportion of variance of shared genetic variant effects (theta): 0.8</span></div><div class='output co'>#&gt; <span class='message'>Proportion of genetic variant effects to have a trait-independent fixed effect (pIndependentGenetic): 0.4</span></div><div class='output co'>#&gt; <span class='message'>Proportion of traits influenced by independent genetic variant effects (pTraitIndependentGenetic): 0.2</span></div><div class='output co'>#&gt; <span class='message'>Proportion of non-linear phenotype transformation (proportionNonlinear): 0</span></div><div class='output co'>#&gt; <span class='message'></span></div><div class='input'>
<span class='co'>#genetic fixed and bg effects</span>
<span class='va'>model</span> <span class='op'>&lt;-</span> <span class='fu'>setModel</span><span class='op'>(</span>genVar<span class='op'>=</span><span class='fl'>1</span>, h2s<span class='op'>=</span><span class='fl'>0.01</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>The total noise variance (noiseVar) is: 0</span></div><div class='output co'>#&gt; <span class='message'>The noise model is: noNoise</span></div><div class='output co'>#&gt; <span class='message'>The total genetic variance (genVar) is: 1</span></div><div class='output co'>#&gt; <span class='message'>The genetic model is: geneticFixedAndBg</span></div><div class='output co'>#&gt; <span class='message'>Proportion of variance of genetic variant effects (h2s): 0.01</span></div><div class='output co'>#&gt; <span class='message'>Proportion of variance of shared genetic variant effects (theta): 0.8</span></div><div class='output co'>#&gt; <span class='message'>Proportion of genetic variant effects to have a trait-independent fixed effect (pIndependentGenetic): 0.4</span></div><div class='output co'>#&gt; <span class='message'>Proportion of traits influenced by independent genetic variant effects (pTraitIndependentGenetic): 0.2</span></div><div class='output co'>#&gt; <span class='message'>Proportion of variance of infinitesimal genetic effects (h2bg): 0.99</span></div><div class='output co'>#&gt; <span class='message'>Proportion of variance of shared infinitesimal genetic effects (eta): 0.8</span></div><div class='output co'>#&gt; <span class='message'>Proportion of non-linear phenotype transformation (proportionNonlinear): 0</span></div><div class='output co'>#&gt; <span class='message'></span></div><div class='input'>
<span class='co'>#genetic and noise fixed effects only</span>
<span class='va'>model</span> <span class='op'>&lt;-</span> <span class='fu'>setModel</span><span class='op'>(</span>genVar<span class='op'>=</span><span class='fl'>0.4</span>, h2s<span class='op'>=</span><span class='fl'>1</span>, delta<span class='op'>=</span><span class='fl'>1</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>The total noise variance (noiseVar) is: 0.6</span></div><div class='output co'>#&gt; <span class='message'>The noise model is: noiseFixedOnly</span></div><div class='output co'>#&gt; <span class='message'>Proportion of non-genetic covariate variance (delta): 1</span></div><div class='output co'>#&gt; <span class='message'>Proportion of variance of shared non-genetic covariate effects (gamma): 0.8</span></div><div class='output co'>#&gt; <span class='message'>Proportion of non-genetic covariates to have a trait-independent effect (pIndependentConfounders ): 0.4</span></div><div class='output co'>#&gt; <span class='message'>Proportion of traits influenced by independent non-genetic covariate effects (pTraitIndependentConfounders): 0.2</span></div><div class='output co'>#&gt; <span class='message'></span></div><div class='output co'>#&gt; <span class='message'>The total genetic variance (genVar) is: 0.4</span></div><div class='output co'>#&gt; <span class='message'>The genetic model is: geneticFixedOnly</span></div><div class='output co'>#&gt; <span class='message'>Proportion of variance of genetic variant effects (h2s): 1</span></div><div class='output co'>#&gt; <span class='message'>Proportion of variance of shared genetic variant effects (theta): 0.8</span></div><div class='output co'>#&gt; <span class='message'>Proportion of genetic variant effects to have a trait-independent fixed effect (pIndependentGenetic): 0.4</span></div><div class='output co'>#&gt; <span class='message'>Proportion of traits influenced by independent genetic variant effects (pTraitIndependentGenetic): 0.2</span></div><div class='output co'>#&gt; <span class='message'>Proportion of non-linear phenotype transformation (proportionNonlinear): 0</span></div><div class='output co'>#&gt; <span class='message'></span></div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Hannah Meyer.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


